site/bps/BP-35-128-bits-support.md

```diff
@@ -0,0 +1,399 @@
+
+###### Metadata Flags
+
+```json
+Metadata: 1 Bytes (Long)
+------------------------
+0x 0 0
+   |__| 
+     |
+ version
+
+----
+Bit 0 - 3: digest type (e.g. CRC32, CRC32C and such)
+Bit 4 - 7: version, the most significant bit of this byte will be always set to 1.
+it will be used for differentiating entry format v1 and v2.
+
+```
+
+We are setting the most significant bit to be `1`. So the first byte in entry v2 will
+be a negative value, which can be used for differentiating entry format v1 and v2.
+The version will be encoded into the first byte. The version will be used for describing
+the entry format.
+
+##### Decoding Entry
+
+The pseudo code for decoding an entry will be described as followings:
+
+```java
+
+ByteBuf entry = ...;
+
+int metadataFlags = entry.getByte();
+
+if (metadataFlags <= 128) { // the entry is encoded in v1 format
+	// decoding the entry in v1 format
+	...
+} else {
+	// decoding the entry in v2 format
+}
+
+```
+
+#### Bookie Storage
+
+##### Journal
+
+A new method should be added in journal `WriteCallback` to handle `ledgerScopeId`.
+
+```java
+public interface WriteCallback {
+
+    void writeComplete(int rc,
+                       long ledgerScopeId,
+                       long ledgerId,
+                       long entryId,
+                       BookieSocketAddress addr,
+                       Object ctx);
+
+    default void writeComplete(int rc,
+                               long ledgerId,
+                               long entryId,
+                               BookieSocketAddress addr,
+                               Object ctx) {
+        writeComplete(rc, 0L, ledgerId, entryId, addr, ctx);
+    }
+
+}
+```
+
+The journal should be changed to be able to retrieve `ledgerScopeId` from the entry
+payload based on [Entry Format](#entry-format).
+
+##### Ledger Storage
+
+###### EntryLogger
+
+1. Methods in `EntryLogger` should be able to accept `ledgerScopeId` as a parameter.
+2. EntryLogger should be updated to retrieve `ledgerScopeId` from the entry payload
+   based on [Entry Format](#entry-format).
+
+###### EntryMemTable
+
+`ledgerScopeId` should be added as part of `EntryKey`.
+
+###### IndexPersistenceMgr
+
+Currently the ledger index files (64 bits) are stored into 2-level-hirechicy
+directories - `<msb-32bits-hex>/<lsb-32bits-hex>/<ledger-id-hex>.idx`.
+
+If `ledgerScopeId` is 0, it will be using existing scheme for storing and retrieving
+ledger index files.
+
+If `ledgerScopeId` is not 0, that means the ledgers are produced by new clients that
+support 128-bits, those ledgers will be stored in a 4-level-hirechicy
+directories -
+`<msb-32bits-hex-ledger-scope-id>/<lsb-32bits-hex-ledger-scope-id>/<msb-32bits-hex-ledger-id>/<lsb-32bits-hex-ledger-id>`.
+
+All the file info caches should be updated to use `<ledgerScopeId, ledgerId>`
+as index keys.
+
+###### IndexInMemPageMgr
+
+The LRU pages map will be updated to use `<ledgerScopeId, ledgerId>` as index
+keys.
+
+###### DBLedgerStorage
```

Comment:
Unless DBLedgerStorage is part of your intended usecase, I would exempt it from this change for now, especially since its primary usecase uses the V2 protocol which is also exempted.

Reply:
I include it for completeness. because DBLedgerStorage is not tied to V2 protocol, it can be used by V3 protocol users.